manticore/core/executor.py

```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
-        self._test_number = manager.Value('i', 0 )
+        # Signals / Callbacks handlers will be invoked potentially at differeent 
+        # worker processes. State provides a local context to save data.
+
+        #Executor
+        self.will_start_run = Signal()
+        self.will_finish_run = Signal()
+        self.will_fork_state = Signal()
+        self.will_backup_state = Signal()
+        self.will_restore_state = Signal()
+        self.will_terminate_state = Signal()
+        self.will_generate_testcase = Signal()
+
+        #Cpu
+        self.will_decode_instruction = Signal()
+        self.will_execute_instruction = Signal()
+        self.did_execute_instruction = Signal()
+        self.will_emulate_instruction = Signal()
+        self.did_emulate_instruction = Signal()
+
+        self.will_read_register = Signal()
+        self.will_write_register = Signal()
+        self.will_read_memory = Signal()
+        self.will_write_memory = Signal()
+
+
+        #Be sure every state will forward us their signals
+        self.will_restore_state += self._register_state_callbacks
+
+        #The main executor lock. Acquire this for accessing shared objects
+        self._lock = manager.Condition(manager.RLock())
+
+        #Shutdown Event
+        self._shutdown = manager.Event()
+
+        #States on storage. Shared dict state name ->  state stats
+        self._states = manager.list()
+
+        #Number of currently running workers. Initially no runnign workers
         self._running = manager.Value('i', 0 )
-        self._count = manager.Value('i', 0 )
-        self._shutdown = manager.Value('i', 0)
-        self._visited = manager.list()
-        self._errors = manager.list()
-        self._all_branches = manager.list()
-        self._lock = manager.Condition(manager.Lock())
-        #Stats
-        self._stats = manager.list()
-
-        #search previously generated states
-        saved_states = [ self._getFilename(filename) for filename in os.listdir(self.workspace) if filename.endswith('.pkl') ]
+
+        #Number of generated testcases
+        self._test_count = manager.Value('i', 0 )
+
+        #Number of total intermediate states
+        self._state_count = manager.Value('i', 0 )
+
+        #Executor wide shared context
+        self._shared_context = manager.dict()
+    
+        #scheduling priority policy (wip)
+        self.policy = Random()
+
+        if not self._load_workspace():
+            state_id = self.backup(initial)
+            self._states.append(state_id)
+
+    @contextmanager
+    def locked_context(self):
+        ''' Executor context is a shared memory object. All workers share this. 
+            It needs a lock. Its used like this:
+
+            with executor.context() as context:
+                vsited = context['visited']
+                visited.append(state.cpu.PC)
+                context['visited'] = visited
+        '''
+        with self._lock:
+            yield self._shared_context
```

Comment:
So _shared_context is shared between all running workers.
If we grant access to one without a lock and the callback at that worker do some non trivial read writes it may lead to a race. no? (there is already an example hook using shared_context at manticore)